app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt

```diff
@@ -1,32 +1,71 @@
 package com.lightricks.feedexercise.ui.feed
 
+import android.app.Application
 import androidx.lifecycle.*
 import com.lightricks.feedexercise.data.FeedItem
+import com.lightricks.feedexercise.data.FeedRepository
+import com.lightricks.feedexercise.database.FeedDatabase
+import com.lightricks.feedexercise.network.FeedApiService
 import com.lightricks.feedexercise.util.Event
+import com.squareup.moshi.Moshi
+import com.squareup.moshi.kotlin.reflect.KotlinJsonAdapterFactory
+import io.reactivex.schedulers.Schedulers
+import okhttp3.OkHttpClient
+import retrofit2.Retrofit
+import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
+import retrofit2.converter.moshi.MoshiConverterFactory
 import java.lang.IllegalArgumentException
 
 /**
  * This view model manages the data for [FeedFragment].
  */
-open class FeedViewModel : ViewModel() {
+open class FeedViewModel(application: Application) : AndroidViewModel(application) {
     private val isLoading = MutableLiveData<Boolean>()
-    private val isEmpty = MutableLiveData<Boolean>()
-    private val feedItems = MediatorLiveData<List<FeedItem>>()
-    private val networkErrorEvent = MutableLiveData<Event<String>>()
+    private val isEmpty: MediatorLiveData<Boolean>
 
+    //    private val feedItems = MediatorLiveData<List<FeedItem>>()
+    private val networkErrorEvent = MutableLiveData<Event<String>>()
+    private var feedRepository: FeedRepository
     fun getIsLoading(): LiveData<Boolean> = isLoading
     fun getIsEmpty(): LiveData<Boolean> = isEmpty
-    fun getFeedItems(): LiveData<List<FeedItem>> = feedItems
+    fun getFeedItems(): LiveData<List<FeedItem>> = feedRepository.getLiveData()
     fun getNetworkErrorEvent(): LiveData<Event<String>> = networkErrorEvent
 
     init {
-        refresh()
+        val feedApiService = initRetrofit()
+        val feedDao = FeedDatabase.getDatabase(this.getApplication()).feedDao()
+        feedRepository = FeedRepository(feedDao, feedApiService)
+        isEmpty = MediatorLiveData<Boolean>().apply {
+            this.addSource(getFeedItems()) {
+                this.postValue(it.isEmpty())
+            }
+        }
     }
 
+
     fun refresh() {
-        //todo: fix the implementation
-        isLoading.value = false
-        isEmpty.value = true
+        isLoading.postValue(true)
+        val completable = feedRepository.refresh()
+            .observeOn(Schedulers.io())
+            .subscribeOn(Schedulers.io())
+            .subscribe(
+                { isLoading.postValue(false) },
+                { networkErrorEvent.postValue(Event<String>(it.message!!)) }
```

Comment:
Two things here:
1. In case of error we also want to hide the loading indicator.
2. Even if the error is `null` we want to show some message. You can write for example:
`Event(it?.message ?: "error")`